### PR TITLE
feat: Firebase

### PR DESCRIPTION
**Pull request by david20571015**
# Description: 

Add services to interact with firebase.

# TODOs:

- [x] unit test
- [x] error handling (will be implemented by [filter](https://docs.nestjs.com/exception-filters))

### CODE REVIEW

File: src/auth/firebase.service.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      editedList,
+      element,
+    );
+    const newSubscribedList = this.upsertElementByArticleId(
+      articleId,
+      subscribedList,
+      element,
+    );
+
+    userRef.set(
```
Comment (sugarbobo-ch) [inline src/auth/firebase.service.ts]: Is this function has callback or is it asynchrony? That we should return promise when success

File: src/auth/firebase.service.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+      { edited: newEditedList, subscribed: newSubscribedList },
```
Comment (sugarbobo-ch) [inline src/auth/firebase.service.ts]: We could use object wrap it and return in LN79

File: src/auth/firebase.service.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      { merge: true },
+    );
+
+    return Promise.resolve({
```
Comment (sugarbobo-ch) [inline src/auth/firebase.service.ts]: asynchrony function default wrap return value in promise resolve, so maybe we don’t need to write in this way.

FYI: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

File: src/auth/firebase.service.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+  async updateArticleSubscription(
+    uid: string,
+    articleId: string,
+    isSubscribe: boolean,
+  ) {
+    const userRef = this.db.collection('articles').doc(uid);
+    const doc = await userRef.get();
+
+    const subscribedList = doc.get('subscribed') ?? [];
```
Comment (sugarbobo-ch) [inline src/auth/firebase.service.ts:99]: Could we write in some readable without “??”

File: src/auth/firebase.service.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+      );
+    } else {
+      newSubscribedList = subscribedList.filter(
+        (s: any) => s.articleId !== articleId,
+      );
+    }
+
+    userRef.set({ subscribed: newSubscribedList }, { merge: true });
+
+    return Promise.resolve(newSubscribedList);
```
Comment (sugarbobo-ch) [inline src/auth/firebase.service.ts]: ditto

File: src/auth/firebase.service.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+    return Promise.resolve(newPoint);
+  }
+
+  async getUserPoints(uid: string) {
+    const userRef = this.db.collection('users').doc(uid);
+    const doc = await userRef.get();
+
+    const point: number = doc.get('points') ?? 0;
+
```
Comment (jinan-tw) [inline src/auth/firebase.service.ts:133]: ditto

File: src/auth/firebase.service.ts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    userRef.set({ point: newPoint }, { merge: true });
+
+    return Promise.resolve(newPoint);
+  }
+
+  async getUserPoints(uid: string) {
+    const userRef = this.db.collection('users').doc(uid);
+    const doc = await userRef.get();
+
+    const point: number = doc.get('points') ?? 0;
```
Comment (sugarbobo-ch) [inline src/auth/firebase.service.ts:139]: ditto

File: src/auth/firebase.service.ts [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+    return Promise.resolve(point);
```
Comment (sugarbobo-ch) [inline src/auth/firebase.service.ts]: ditto

File: src/auth/firebase.service.ts [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+    return Promise.resolve(newNotiList);
+  }
+
+  async getNotifications(uid: string) {
+    const notiRef = this.db.collection('notifications').doc(uid);
+    const doc = await notiRef.get();
+    const notiList = doc.get('notifications') ?? [];
+
+    return Promise.resolve(notiList);
```
Comment (sugarbobo-ch) [inline src/auth/firebase.service.ts]: ditto

File: src/auth/firebase.service.ts [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+  async updateNotification(uid: string, notifications: any[], maxRemain = 20) {
+    const notiRef = this.db.collection('notifications').doc(uid);
+    const doc = await notiRef.get();
+    const notiList = doc.get('notifications') ?? [];
+
+    const newNotiList = [...notiList, ...notifications].slice(-maxRemain);
+
+    notiRef.set({ notifications: newNotiList }, { merge: true });
+
+    return Promise.resolve(newNotiList);
```
Comment (sugarbobo-ch) [inline src/auth/firebase.service.ts]: ditto

File: src/auth/firebase.service.ts [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+  async getUserPoints(uid: string) {
+    const userRef = this.db.collection('users').doc(uid);
+    const doc = await userRef.get();
+
+    const point: number = doc.get('points') ?? 0;
+
+    return Promise.resolve(point);
+  }
+
+  async updateNotification(uid: string, notifications: any[], maxRemain = 20) {
```
Comment (sugarbobo-ch) [inline src/auth/firebase.service.ts]: maxRemain should write in const value as DEFAULT_MAX_REMAIN on the top of this file.

And does notification has its type? Not good in use any as type

File: src/auth/firebase.service.ts [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    return this.db.collection('users').doc(uid).get();
+  }
+
+  private upsertElementByArticleId = (
+    articleId: string,
+    list: any[],
+    element: any,
+  ) => {
+    const newArray = [...list];
+    const index = list.findIndex((s) => s.articleId === articleId);
```
Comment (sugarbobo-ch) [inline src/auth/firebase.service.ts:57]: We could use list.find(…) ?newArray.push(element): …

File: src/auth/firebase.service.ts (hunk 4 above)
Comment (david20571015) [reply src/auth/firebase.service.ts:99]: Do you mean explicitly check whether it is unknown by `if` statement?

File: src/auth/firebase.service.ts (hunk 12 above)
Comment (david20571015) [reply src/auth/firebase.service.ts:57]: The index is still needed to modify the element if found. Is there a better practice to do the same thing?

File: src/auth/firebase.service.ts (hunk 4 above)
Comment (sugarbobo-ch) [reply src/auth/firebase.service.ts:99]: yup

File: src/auth/firebase.service.ts (hunk 12 above)
Comment (sugarbobo-ch) [reply src/auth/firebase.service.ts:57]: Try this:
```
const newArray = list.reduce((acc, curr) => {
    if (curr.articleId === element.articleId) {
      return [...acc, element];
    } else {
      return [...acc, curr];
    }
  }, []);

  if (!newArray.includes(element)) {
    newArray.push(element);
  }

  return newArray;
```

Comment (sugarbobo-ch) [reply src/auth/firebase.service.ts:57]: Your answer is same to ChatGPT, so just choose what u want, the perfeormance are simlliar
image

File: src/auth/firebase.service.ts (hunk 2 above)
Comment (david20571015) [reply src/auth/firebase.service.ts]: done.

File: src/auth/firebase.service.ts (hunk 3 above)
Comment (david20571015) [reply src/auth/firebase.service.ts]: done.

File: src/auth/firebase.service.ts (hunk 5 above)
Comment (david20571015) [reply src/auth/firebase.service.ts]: done.

File: src/auth/firebase.service.ts (hunk 7 above)
Comment (david20571015) [reply src/auth/firebase.service.ts:139]: done.

File: src/auth/firebase.service.ts (hunk 8 above)
Comment (david20571015) [reply src/auth/firebase.service.ts]: done.

File: src/auth/firebase.service.ts (hunk 9 above)
Comment (david20571015) [reply src/auth/firebase.service.ts]: done.

File: src/auth/firebase.service.ts (hunk 10 above)
Comment (david20571015) [reply src/auth/firebase.service.ts]: done.

File: src/auth/firebase.service.ts (hunk 4 above)
Comment (david20571015) [reply src/auth/firebase.service.ts:99]: In this case, `doc.get('subscribed')` returns `undefined` if no such field exists.
So, I think using nullish coalescing operator is more concise and more readable.

File: src/auth/firebase.service.ts (hunk 11 above)
Comment (david20571015) [reply src/auth/firebase.service.ts]: done.

File: src/auth/firebase.service.ts (hunk 12 above)
Comment (david20571015) [reply src/auth/firebase.service.ts:57]: I choose to remain my implementation.

File: src/auth/firebase.service.ts (hunk 1 above)
Comment (david20571015) [reply src/auth/firebase.service.ts]: done.
I use `await` for awaitable functions to catch the errors.
`set` only returns the write time when success, so I return a promise with written datas as [previous implementation](https://github.com/NCTU-SYNC/sync-server/blob/81a349a95954442447014f2855cc29db95369337/src/utils/firebase.js#L66).

**Pull request merged**